crates/samples/client/src/main.rs

```diff
@@ -3,47 +3,93 @@
-use mssf_core::sync::wait::WaitableCallback;
+use std::time::Duration;
 
-fn main() -> mssf_core::Result<()> {
-    println!("GetNodeCli");
+use mssf_core::{
+    WString,
+    client::{ClaimsRetrievalMetadata, FabricClient, GatewayInformationResult},
+    types::{FabricClaimsCredentials, Uri},
+};
 
-    let c: IFabricQueryClient = mssf_core::API_TABLE
-        .fabric_create_local_client3::<IFabricQueryClient>(None, None)
-        .expect("cannot get localclient");
+// Sample client that connects to a Service Fabric cluster with AAD enabled.
+// This is similar to how SF pwsh client connects to such cluster.
+// Run with: cargo run --bin samples_client -- <host:port>
+// This is tested against with a AAD enabled cluster.
+// TODO: expand this sample to parse args and do some actual requests.
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    tracing_subscriber::fmt().init();
 
-    let (token, callback) = WaitableCallback::channel();
+    // get endpoint from argv
+    let args: Vec<String> = std::env::args().collect();
+    assert!(args.len() == 2, "Usage: client <host:port>");
+    let env_endpoint = WString::from(&args[1]);
 
-    let querydescription = FABRIC_NODE_QUERY_DESCRIPTION::default();
+    // split host and port
+    let endpoint_str = env_endpoint.to_string_lossy();
+    let mut parts = endpoint_str.split(':');
+    let host_str = parts.next().unwrap();
+    assert_eq!(parts.next().unwrap(), "19000");
+    tracing::info!("Using host_str: {}", host_str);
 
-    let ctx = unsafe {
-        c.BeginGetNodeList(&querydescription, 1000, &callback)
-            .expect("cannot get ctx")
-    };
+    // channel for client connection notification
+    let (cc_tx, mut cc_rx) = tokio::sync::mpsc::channel::<GatewayInformationResult>(1);
+    // channel for message retrieval notification
+    let (claims_tx, mut claims_rx) = tokio::sync::mpsc::channel::<ClaimsRetrievalMetadata>(1);
+    let fc = FabricClient::builder()
+        .with_connection_strings(vec![env_endpoint])
+        .with_on_client_connect(move |gw| {
+            tracing::info!("Client connected: {:?}", gw);
+            cc_tx.blocking_send(gw.clone()).expect("cannot send");
+            Ok(())
+        })
+        .with_on_client_disconnect(move |_| {
+            // This is not invoked in this test. FabricClient does not invoke this on drop.
+            panic!("client disconnected");
+        })
+        .with_on_claims_retrieval(move |meta| {
+            claims_tx
+                .blocking_send(meta.clone())
+                .expect("cannot send claims retrieval metadata");
+            // For test purpose, return empty claims.
+            // Empty claims will trigger default handler to run.
+            // See: https://github.com/microsoft/service-fabric/blob/36f7531df0fd990f8af1792ae2cd5cf811521ab3/src/prod/src/client/ClientConnectionManager.cpp#L933
+            // Ok(WString::from("_Invalid_"))
```

Comment:
Change this back.
There is a problem invoking the default claims retrieval handler due to dotnet dll loading.